Wrappers/Python/cil/framework/framework.py

```diff
@@ -40,6 +40,205 @@
+            boolbatches.append(boolbatch)
+        
+        return boolbatches
+
+    def partition(self, num_batches, mode, seed=None):
+        '''Partition the data into ``num_batches`` batches using the specified ``mode``.
+        
+
+        The modes are
+        
+        1. ``sequential`` - The data will be partitioned into ``num_batches`` batches of sequential indices.
+        
+        2. ``staggered`` - The data will be partitioned into ``num_batches`` batches of sequential indices, with stride equal to ``num_batches``.
+        
+        3. ``random_permutation`` - The data will be partitioned into ``num_batches`` batches of random indices.
+
+        Parameters
+        ----------
+            num_batches : int
+                The number of batches to partition the data into.
+            mode : str
+                The mode to use for partitioning. Must be one of ``sequential``, ``staggered`` or ``random_permutation``.
+            seed : int, optional
+                The seed to use for the random permutation. If not specified, the random number
+                generator will not be seeded.
+
+
+        Returns
+        -------
+            BlockDataContainer
+
+        Example
+        -------
+        
+        Partitioning a list of ints [0, 1, 2, 3, 4, 5, 6, 7, 8] into 4 batches will return:
+    
+        1. [[0, 1, 2], [3, 4], [5, 6], [7, 8]] with ``sequential``
+        2. [[0, 4, 8], [1, 5], [2, 6], [3, 7]] with ``staggered``
+        3. [[8, 2, 6], [7, 1], [0, 4], [3, 5]] with ``random_permutation`` and seed 1
+
+        '''
+        if mode == Partitioner.SEQUENTIAL:
+            return self._partition_deterministic(num_batches, stagger=False)
+        elif mode == Partitioner.STAGGERED:
+            return self._partition_deterministic(num_batches, stagger=True)
+        elif mode == Partitioner.RANDOM_PERMUTATION:
+            return self._partition_random_permutation(num_batches, seed=seed)
+        else:
+            raise ValueError('Unknown partition mode {}'.format(mode))
+
+    def _partition_deterministic(self, num_batches, stagger=False, indices=None):
+        '''Partition the data into ``num_batches`` batches.
+        
+        Parameters
+        ----------
+        num_batches : int
+            The number of batches to partition the data into.
+        stagger : bool, optional
+            If ``True``, the batches will be staggered. Default is ``False``.
+        indices : list of int, optional
+            The indices to partition. If not specified, the indices will be generated from the number of projections.
+        '''
+        if indices is None:
+            indices = self.geometry.num_projections
+        masks = self._partition_indices(num_batches, indices, stagger)
+        masks = self._convert_indices_to_masks(masks, indices)
+        blk_geo = self._convert_masks_to_BlockGeometry(masks)
+        
+        # copy data
+        out = blk_geo.allocate(None)
+        out.geometry = blk_geo
```

Comment:
doesn't `allocate` handle this?

Reply:
no, in `BlockDataContainer` we don't enforce the `geometry` field to be present.